contracts/ibc-callbacks/src/contract.rs

```diff
@@ -0,0 +1,114 @@
+use cosmwasm_std::{
+    entry_point, to_json_binary, to_json_string, Binary, Deps, DepsMut, Empty, Env,
+    IbcBasicResponse, IbcCallbackData, IbcDestinationChainCallbackMsg, IbcDstCallback, IbcMsg,
+    IbcSourceChainCallbackMsg, IbcSrcCallback, IbcTimeout, MessageInfo, Response, StdError,
+    StdResult,
+};
+
+use crate::msg::{CallbackType, ExecuteMsg, QueryMsg};
+use crate::state::{load_stats, save_stats, CallbackStats};
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    _msg: Empty,
+) -> StdResult<Response> {
+    // initialize counts
+    let counts = CallbackStats::default();
+    save_stats(deps.storage, &counts)?;
+
+    Ok(Response::new().add_attribute("action", "instantiate"))
+}
+
+/// Sends an `IbcMsg::Transfer` to the given `to_address` on the given `channel_id`.
+#[entry_point]
+pub fn execute(
+    _deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> StdResult<Response> {
+    let src_callback = IbcSrcCallback {
+        address: env.contract.address,
+        gas_limit: None,
+    };
+    let dst_callback = IbcDstCallback {
+        address: msg.to_address.clone(),
+        gas_limit: None,
+    };
+    let transfer_msg = match &*info.funds {
+        [coin] if !coin.amount.is_zero() => IbcMsg::Transfer {
+            to_address: msg.to_address,
+            timeout: IbcTimeout::with_timestamp(
+                env.block.time.plus_seconds(msg.timeout_seconds as u64),
+            ),
+            channel_id: msg.channel_id,
+            amount: coin.clone(),
+            memo: Some(to_json_string(&match msg.callback_type {
```

Comment:
You could create an interesting API if you allow something like `transfer_msg.with_src_callback(env, gas_limit)` instead of writing it to the memo directly here. The method `.with_src_callback` would also return an error if the memo is not empty to ensure we are not overriding important info.

Reply:
That looks interesting indeed. I think it would be even better if we encode this in the type system instead of using runtime errors. We would provide something like a `TransferMsgBuilder` that takes everything except the memo and provides `with_src_callback`, `with_callbacks`, `with_memo`, etc. to convert to a TransferMsg. That way you cannot misuse it accidentally.

Reply:
Yes. You could leave it as is for now and make a followup PR if you think this needs further design

Reply:
Added a new issue #2122 for this.